lib/internal/fetch/headers.js

```diff
@@ -0,0 +1,225 @@
+ * iterating at most half the length of the input array. The search also
+ * assumes all entries are strings and uses String.prototype.localeCompare for
+ * comparison.
+ */
+function binarySearch(arr, val) {
+  let low = 0;
+  let high = MathFloor(arr.length / 2);
+
+  while (high > low) {
+    const mid = (high + low) >>> 1;
+
+    if (StringPrototypeLocaleCompare(val, arr[mid * 2]) > 0) {
+      low = mid + 1;
+    } else {
+      high = mid;
+    }
+  }
+
+  return low * 2;
+}
+
+function normalizeAndValidateHeaderName(name) {
+  if (name === undefined) {
+    throw new ERR_INVALID_HTTP_TOKEN('Header name', name);
+  }
+  const normalizedHeaderName = StringPrototypeToLocaleLowerCase(name);
+  validateHeaderName(normalizedHeaderName);
+  return normalizedHeaderName;
+}
+
+function normalizeAndValidateHeaderValue(name, value) {
+  if (value === undefined) {
+    throw new ERR_HTTP_INVALID_HEADER_VALUE(value, name);
+  }
+  // https://fetch.spec.whatwg.org/#concept-header-value-normalize
+  const normalizedHeaderValue = RegExpPrototypeSymbolReplace(
+    /^[\n\t\r\x20]+|[\n\t\r\x20]+$/g,
+    value,
+    ''
+  );
+  validateHeaderValue(name, normalizedHeaderValue);
+  return normalizedHeaderValue;
+}
+
+function fill(headers, object) {
+  if (kHeadersList in object) {
+    // Object is instance of Headers
+    headers[kHeadersList] = ArrayPrototypeSlice(object[kHeadersList]);
+  } else if (ArrayIsArray(object)) {
+    // Support both 1D and 2D arrays of header entries
+    if (ArrayIsArray(object[0])) {
+      // Array of arrays
+      for (let i = 0; i < object.length; i++) {
+        if (object[i].length !== 2) {
+          throw new ERR_INVALID_ARG_VALUE('init', header,
+                                          'is not of length 2');
+        }
+        headers.append(object[i][0], object[i][1]);
+      }
+    } else if (typeof object[0] === 'string' || Buffer.isBuffer(object[0])) {
+      // Flat array of strings or Buffers
+      if (object.length % 2 !== 0) {
+        throw new ERR_INVALID_ARG_VALUE('init', object,
+                                        'is not even in length');
+      }
+      for (let i = 0; i < object.length; i += 2) {
+        headers.append(
+          object[i].toString('utf-8'),
+          object[i + 1].toString('utf-8')
+        );
+      }
+    } else {
+      // All other array based entries
+      throw new ERR_INVALID_ARG_VALUE(
+        'init',
+        object,
+        'is not a valid array entry'
+      );
+    }
+  } else if (!isBoxedPrimitive(object)) {
+    // Object of key/value entries
+    const entries = ObjectEntries(object);
+    for (let i = 0; i < entries.length; i++) {
+      headers.append(entries[i][0], entries[i][1]);
+    }
+  }
+}
+
+class Headers {
+  constructor(init) {
+    this[kHeadersList] = [];
+
+    if (init &&
+      validateObject(init, 'init', { allowArray: true }) === undefined) {
+      fill(this, init);
+    }
+  }
```

Comment:
I would simplify this. Default the value of `init` to `{}` and just call `validateObject()` -- `validateObject()` does not return any value so checking the return with `=== undefined` is not necessary.

```suggestion
  constructor(init = {}) {
    validateObject(init, 'init', { allowArray: true });
    this[kHeadersList] = [];
    fill(this, init);
  }
```